Modules/_io/textio.c

```diff
@@ -1719,21 +1719,31 @@ _io_TextIOWrapper_write_impl(textio *self, PyObject *text)
         bytes_len = PyBytes_GET_SIZE(b);
     }
 
-    if (self->pending_bytes == NULL) {
-        self->pending_bytes_count = 0;
-        self->pending_bytes = b;
-    }
-    else if (self->pending_bytes_count + bytes_len > self->chunk_size) {
+    PyObject *pending = self->pending_bytes;
+
+    if (self->pending_bytes_count + bytes_len > self->chunk_size) {
         // Prevent to concatenate more than chunk_size data.
         if (_textiowrapper_writeflush(self) < 0) {
             Py_DECREF(b);
             return NULL;
         }
+        if (self->pending_bytes) {
+            // gh-119506: call to _textiowrapper_writeflush()
+            // can write new data to pending_bytes
+            pending = self->pending_bytes;
+            self->pending_bytes = b;
+            b = pending;
+        }
```

Comment:
why do you swap data here?
* `self->pending` can be list
* call from other thread might be returned already.
  * "first returned first write" seems correct than "first called first write".

Reply:
If you look at the test case, where flushing the stream triggers a write, it's not a threading thing, it's because of write being called recursively.

Reply:
Even single thread case, it is not possble to guarantee "first called, first written."
Simple implementation is more important.